src/main/java/bio/terra/cda/app/controller/QueryApiController.java

```diff
@@ -18,58 +20,79 @@
 @Controller
 public class QueryApiController implements QueryApi {
 
-  public static final int DEFAULT_LIMIT = 100;
-  public static final int DEFAULT_OFFSET = 0;
+  public static final int DEFAULT_PAGE_SIZE = 100;
+  public static final int DEFAULT_OFFSET = 100;
 
   private final QueryService queryService;
   private final ApplicationConfiguration applicationConfiguration;
+  private final HttpServletRequest webRequest;
 
   @Autowired
   public QueryApiController(
-      QueryService queryService, ApplicationConfiguration applicationConfiguration) {
+      QueryService queryService,
+      ApplicationConfiguration applicationConfiguration,
+      HttpServletRequest webRequest) {
     this.queryService = queryService;
     this.applicationConfiguration = applicationConfiguration;
+    this.webRequest = webRequest;
   }
 
-  private ResponseEntity<QueryResponseData> sendQuery(
-      @Valid String querySql, @Valid Integer offset, @Valid Integer limit, boolean dryRun) {
-    String queryStringWithPagination =
-        String.format(
-            "%s LIMIT %d OFFSET %d",
-            querySql,
-            Objects.requireNonNullElse(limit, DEFAULT_LIMIT),
-            Objects.requireNonNullElse(offset, DEFAULT_OFFSET));
+  private String createNextUrl(String jobId, int offset, int pageSize) {
+    var path = String.format("/api/v1/query/%s?offset=%s&pageSize=%s", jobId, offset, pageSize);
 
-    var result =
-        dryRun ? Collections.emptyList() : queryService.runQuery(queryStringWithPagination);
-    var response = new QueryResponseData().result(new ArrayList<>(result)).querySql(querySql);
+    try {
+      URL baseUrl = new URL(webRequest.getHeader("referer"));
+      return new URL(baseUrl.getProtocol(), baseUrl.getHost(), baseUrl.getPort(), path).toString();
+    } catch (MalformedURLException e) {
+      // Not sure what a good fallback would be here.
+      return path;
+    }
+  }
+
+  @Override
+  public ResponseEntity<QueryResponseData> query(
+      String id, Integer boxedOffset, Integer boxedPageSize) {
+    int offset = boxedOffset != null ? boxedOffset : DEFAULT_OFFSET;
+    int pageSize = boxedPageSize != null ? boxedPageSize : DEFAULT_PAGE_SIZE;
+    var result = queryService.getQueryResults(id, offset, pageSize);
+    var response =
+        new QueryResponseData()
+            .result(new ArrayList<>(result.items))
```

Comment:
is `result.items` not a list?  If it's an array, maybe `List.of(result.item)`?

Reply:
It's a list but it's of `List<JsonNode>` but the openapi generated Model code uses `List<Object>`. AFAIK there's no great way to get from one type to the other.

Reply:
The internet says using `Collections.unmodifiableList()` is slightly better, I can update it to use that.